---
title: React ç»„ä»¶å¼€å‘è§„èŒƒ
type: context_aware
priority: high
description: React ç»„ä»¶çš„å¼€å‘è§„èŒƒå’Œæœ€ä½³å®è·µ
enabled: true
tags:
  - react
  - components
  - frontend
triggers:
  fileExtensions:
    - .tsx
    - .jsx
  pathPatterns:
    - src/components/**
    - src/pages/**
    - packages/*/src/components/**
    - webview/src/components/**
---

# React ç»„ä»¶å¼€å‘è§„èŒƒ

## ç»„ä»¶ç»“æ„

### å‡½æ•°ç»„ä»¶æ¨¡æ¿

âœ… **æ¨èçš„ç»„ä»¶ç»“æ„**:

```typescript
import React, { useState, useEffect, useCallback } from 'react';
import './MyComponent.css';

interface MyComponentProps {
  title: string;
  onSave?: (data: Data) => void;
  className?: string;
}

export const MyComponent: React.FC<MyComponentProps> = ({
  title,
  onSave,
  className
}) => {
  // ğŸ¯ 1. State å£°æ˜
  const [data, setData] = useState<Data | null>(null);
  const [loading, setLoading] = useState(false);

  // ğŸ¯ 2. Refs
  const inputRef = useRef<HTMLInputElement>(null);

  // ğŸ¯ 3. è‡ªå®šä¹‰ Hooks
  const { user } = useAuth();

  // ğŸ¯ 4. useEffectï¼ˆå‰¯ä½œç”¨ï¼‰
  useEffect(() => {
    loadData();
    return () => cleanup();
  }, []);

  // ğŸ¯ 5. äº‹ä»¶å¤„ç†å‡½æ•°
  const handleSave = useCallback(() => {
    if (data) {
      onSave?.(data);
    }
  }, [data, onSave]);

  // ğŸ¯ 6. è¾…åŠ©å‡½æ•°
  const loadData = async () => {
    setLoading(true);
    try {
      const result = await fetchData();
      setData(result);
    } finally {
      setLoading(false);
    }
  };

  // ğŸ¯ 7. æ¡ä»¶æ¸²æŸ“
  if (loading) {
    return <LoadingSpinner />;
  }

  if (!data) {
    return <EmptyState />;
  }

  // ğŸ¯ 8. ä¸»æ¸²æŸ“
  return (
    <div className={`my-component ${className || ''}`}>
      <h2>{title}</h2>
      <DataDisplay data={data} />
      <button onClick={handleSave}>Save</button>
    </div>
  );
};
```

## Props å®šä¹‰

### å‘½åè§„èŒƒ

- Props æ¥å£å‘½åï¼š`{ComponentName}Props`
- å¯é€‰ props ä½¿ç”¨ `?`
- å›è°ƒå‡½æ•°ä»¥ `on` å¼€å¤´

```typescript
interface UserCardProps {
  user: User;
  editable?: boolean;
  onEdit?: (user: User) => void;
  onDelete?: (userId: string) => void;
  className?: string;
  style?: React.CSSProperties;
}
```

### é»˜è®¤å€¼

ä½¿ç”¨è§£æ„èµ‹å€¼è®¾ç½®é»˜è®¤å€¼ï¼š

```typescript
export const UserCard: React.FC<UserCardProps> = ({
  user,
  editable = false,
  onEdit,
  className = '',
}) => {
  // ...
};
```

### Children Props

æ­£ç¡®å¤„ç† childrenï¼š

```typescript
interface ModalProps {
  isOpen: boolean;
  onClose: () => void;
  children: React.ReactNode;
}

export const Modal: React.FC<ModalProps> = ({ isOpen, onClose, children }) => {
  if (!isOpen) return null;

  return (
    <div className="modal-overlay" onClick={onClose}>
      <div className="modal-content" onClick={(e) => e.stopPropagation()}>
        {children}
      </div>
    </div>
  );
};
```

## Hooks ä½¿ç”¨

### useState

- ä½¿ç”¨æè¿°æ€§çš„çŠ¶æ€åç§°
- å¤æ‚çŠ¶æ€ä½¿ç”¨ `useReducer`

```typescript
// âœ… å¥½çš„å‘½å
const [isLoading, setIsLoading] = useState(false);
const [userData, setUserData] = useState<User | null>(null);
const [formErrors, setFormErrors] = useState<Record<string, string>>({});

// âŒ ä¸å¥½çš„å‘½å
const [flag, setFlag] = useState(false);
const [data, setData] = useState(null);
```

### useEffect

- æ˜ç¡®ä¾èµ–é¡¹æ•°ç»„
- è¿”å›æ¸…ç†å‡½æ•°ï¼ˆå¦‚æœéœ€è¦ï¼‰
- æ‹†åˆ†ä¸ç›¸å…³çš„å‰¯ä½œç”¨

âœ… **æ¨è**:
```typescript
// ç‹¬ç«‹çš„å‰¯ä½œç”¨
useEffect(() => {
  document.title = `User: ${user.name}`;
}, [user.name]);

useEffect(() => {
  const timer = setInterval(() => {
    checkStatus();
  }, 5000);

  return () => clearInterval(timer);
}, []);
```

âŒ **ä¸æ¨è**:
```typescript
// æ··åˆå¤šä¸ªä¸ç›¸å…³çš„å‰¯ä½œç”¨
useEffect(() => {
  document.title = `User: ${user.name}`;
  const timer = setInterval(() => checkStatus(), 5000);
  return () => clearInterval(timer);
}, [user.name]);
```

### useCallback å’Œ useMemo

ä»…åœ¨å¿…è¦æ—¶ä½¿ç”¨ï¼š

```typescript
// âœ… å¥½çš„ç”¨æ³• - ä¼ é€’ç»™å­ç»„ä»¶çš„å›è°ƒ
const handleClick = useCallback(() => {
  processData(data);
}, [data]);

<ExpensiveChild onClick={handleClick} />

// âœ… å¥½çš„ç”¨æ³• - æ˜‚è´µçš„è®¡ç®—
const sortedItems = useMemo(
  () => items.sort((a, b) => heavyComparison(a, b)),
  [items]
);

// âŒ ä¸å¿…è¦çš„ä¼˜åŒ–
const simpleValue = useMemo(() => x + y, [x, y]);
```

### è‡ªå®šä¹‰ Hooks

æå–å¯å¤ç”¨é€»è¾‘åˆ°è‡ªå®šä¹‰ Hooksï¼š

```typescript
function useLocalStorage<T>(key: string, initialValue: T) {
  const [storedValue, setStoredValue] = useState<T>(() => {
    try {
      const item = window.localStorage.getItem(key);
      return item ? JSON.parse(item) : initialValue;
    } catch (error) {
      console.error(error);
      return initialValue;
    }
  });

  const setValue = useCallback((value: T | ((val: T) => T)) => {
    try {
      const valueToStore = value instanceof Function ? value(storedValue) : value;
      setStoredValue(valueToStore);
      window.localStorage.setItem(key, JSON.stringify(valueToStore));
    } catch (error) {
      console.error(error);
    }
  }, [key, storedValue]);

  return [storedValue, setValue] as const;
}

// ä½¿ç”¨
const [user, setUser] = useLocalStorage<User>('user', defaultUser);
```

## æ¡ä»¶æ¸²æŸ“

### çŸ­è·¯è¿ç®—ç¬¦

```typescript
// âœ… ç®€å•æ¡ä»¶
{isLoggedIn && <UserProfile />}
{error && <ErrorMessage error={error} />}

// âš ï¸ æ³¨æ„ï¼š0 å’Œç©ºå­—ç¬¦ä¸²ä¼šè¢«æ¸²æŸ“
{count && <span>{count} items</span>}  // å½“ count=0 æ—¶ä¼šæ˜¾ç¤º "0"

// âœ… æ­£ç¡®åšæ³•
{count > 0 && <span>{count} items</span>}
```

### ä¸‰å…ƒè¿ç®—ç¬¦

```typescript
// âœ… ç®€å•äºŒé€‰ä¸€
{isLoading ? <Spinner /> : <Content />}

// âŒ åµŒå¥—è¿‡æ·±
{isLoading ? (
  <Spinner />
) : hasError ? (
  <Error />
) : hasData ? (
  <Content data={data} />
) : (
  <Empty />
)}

// âœ… æå‰è¿”å›æˆ–ä½¿ç”¨è¾…åŠ©å‡½æ•°
const renderContent = () => {
  if (isLoading) return <Spinner />;
  if (hasError) return <Error />;
  if (!hasData) return <Empty />;
  return <Content data={data} />;
};

return <div>{renderContent()}</div>;
```

## åˆ—è¡¨æ¸²æŸ“

### Key çš„ä½¿ç”¨

```typescript
// âœ… ä½¿ç”¨ç¨³å®šçš„å”¯ä¸€ ID
{users.map(user => (
  <UserCard key={user.id} user={user} />
))}

// âŒ ä½¿ç”¨ç´¢å¼•ï¼ˆé™¤éåˆ—è¡¨é™æ€ä¸”ä¸ä¼šé‡æ’åºï¼‰
{users.map((user, index) => (
  <UserCard key={index} user={user} />
))}
```

### æ€§èƒ½ä¼˜åŒ–

```typescript
// âœ… ä½¿ç”¨ React.memo é¿å…ä¸å¿…è¦çš„é‡æ¸²æŸ“
export const UserCard = React.memo<UserCardProps>(({ user, onEdit }) => {
  return (
    <div className="user-card">
      <h3>{user.name}</h3>
      <button onClick={() => onEdit(user)}>Edit</button>
    </div>
  );
});

// è‡ªå®šä¹‰æ¯”è¾ƒå‡½æ•°
export const UserCard = React.memo<UserCardProps>(
  ({ user, onEdit }) => { /* ... */ },
  (prevProps, nextProps) => {
    return prevProps.user.id === nextProps.user.id;
  }
);
```

## æ ·å¼

### CSS Modules

ä¼˜å…ˆä½¿ç”¨ CSS Modulesï¼š

```typescript
import styles from './MyComponent.module.css';

export const MyComponent: React.FC = () => {
  return (
    <div className={styles.container}>
      <h2 className={styles.title}>Title</h2>
    </div>
  );
};
```

### æ¡ä»¶ç±»å

ä½¿ç”¨ `classnames` åº“æˆ–æ¨¡æ¿å­—ç¬¦ä¸²ï¼š

```typescript
import classNames from 'classnames';

const buttonClass = classNames(
  styles.button,
  {
    [styles.active]: isActive,
    [styles.disabled]: isDisabled,
  },
  className
);

// æˆ–ä½¿ç”¨æ¨¡æ¿å­—ç¬¦ä¸²
const buttonClass = `${styles.button} ${isActive ? styles.active : ''} ${className || ''}`;
```

## è¡¨å•å¤„ç†

### å—æ§ç»„ä»¶

```typescript
export const UserForm: React.FC<UserFormProps> = ({ onSubmit }) => {
  const [formData, setFormData] = useState({
    name: '',
    email: '',
  });

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;
    setFormData(prev => ({ ...prev, [name]: value }));
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    onSubmit(formData);
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        name="name"
        value={formData.name}
        onChange={handleChange}
        placeholder="Name"
      />
      <input
        name="email"
        type="email"
        value={formData.email}
        onChange={handleChange}
        placeholder="Email"
      />
      <button type="submit">Submit</button>
    </form>
  );
};
```

### è¡¨å•éªŒè¯

```typescript
interface FormErrors {
  [key: string]: string;
}

export const UserForm: React.FC = () => {
  const [formData, setFormData] = useState({ name: '', email: '' });
  const [errors, setErrors] = useState<FormErrors>({});

  const validate = (): boolean => {
    const newErrors: FormErrors = {};

    if (!formData.name.trim()) {
      newErrors.name = 'Name is required';
    }

    if (!formData.email.includes('@')) {
      newErrors.email = 'Invalid email format';
    }

    setErrors(newErrors);
    return Object.keys(newErrors).length === 0;
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (validate()) {
      // Submit form
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <input name="name" value={formData.name} onChange={handleChange} />
        {errors.name && <span className="error">{errors.name}</span>}
      </div>
      {/* ... */}
    </form>
  );
};
```

## é”™è¯¯å¤„ç†

### Error Boundary

åˆ›å»ºé”™è¯¯è¾¹ç•Œç»„ä»¶ï¼š

```typescript
interface ErrorBoundaryProps {
  children: React.ReactNode;
  fallback?: React.ReactNode;
}

interface ErrorBoundaryState {
  hasError: boolean;
  error?: Error;
}

export class ErrorBoundary extends React.Component<
  ErrorBoundaryProps,
  ErrorBoundaryState
> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    return { hasError: true, error };
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    console.error('Error caught by boundary:', error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      return this.props.fallback || (
        <div className="error-fallback">
          <h2>Something went wrong</h2>
          <p>{this.state.error?.message}</p>
        </div>
      );
    }

    return this.props.children;
  }
}

// ä½¿ç”¨
<ErrorBoundary fallback={<ErrorFallback />}>
  <MyComponent />
</ErrorBoundary>
```

## æµ‹è¯•

### ç»„ä»¶æµ‹è¯•æ¨¡æ¿

```typescript
import { render, screen, fireEvent } from '@testing-library/react';
import { UserCard } from './UserCard';

describe('UserCard', () => {
  const mockUser = {
    id: '1',
    name: 'John Doe',
    email: 'john@example.com',
  };

  it('renders user information', () => {
    render(<UserCard user={mockUser} />);

    expect(screen.getByText(mockUser.name)).toBeInTheDocument();
    expect(screen.getByText(mockUser.email)).toBeInTheDocument();
  });

  it('calls onEdit when edit button is clicked', () => {
    const handleEdit = jest.fn();
    render(<UserCard user={mockUser} onEdit={handleEdit} />);

    fireEvent.click(screen.getByRole('button', { name: /edit/i }));

    expect(handleEdit).toHaveBeenCalledWith(mockUser);
  });
});
```

## å¯è®¿é—®æ€§ (a11y)

### è¯­ä¹‰åŒ– HTML

```typescript
// âœ… å¥½çš„åšæ³•
<button onClick={handleClick}>Click me</button>
<nav><ul><li><a href="/home">Home</a></li></ul></nav>

// âŒ ä¸å¥½çš„åšæ³•
<div onClick={handleClick}>Click me</div>
<div className="nav"><div className="menu">...</div></div>
```

### ARIA å±æ€§

```typescript
<button
  onClick={handleToggle}
  aria-label="Toggle menu"
  aria-expanded={isOpen}
  aria-controls="menu-items"
>
  Menu
</button>

<div id="menu-items" role="menu" hidden={!isOpen}>
  {/* Menu items */}
</div>
```

---

éµå¾ªè¿™äº› React ç»„ä»¶å¼€å‘è§„èŒƒå¯ä»¥æé«˜ä»£ç è´¨é‡ã€å¯ç»´æŠ¤æ€§å’Œå›¢é˜Ÿåä½œæ•ˆç‡ã€‚
